force-app/main/default/classes/SOQL.cls

```diff
@@ -1,361 +1,585 @@
+            false,
+            true,
+            false,
+            false
+          );
+          oppLineItem.quantity += randNum;
+          oppLineItem.PricebookEntryId = oli.PricebookEntryId;
+          oppLineItem.OpportunityId = oppId;
+          newLineItemLst.add(oppLineItem);
+        }
+      }
+      // Insert the new line items
+      insert newLineItemLst;
     }
-    
-    /**
-     * Question 24
-     * Retrieves the Opportunity Name, Account Name, and Owner Name for Opportunities that have a Stage of 'Closed Won'.
-     * Example: getOpportunityDetailsForClosedWon()
-     * @return A list of filtered Opportunity
-     */
-    public static List<Opportunity> getOpportunityDetailsForClosedWon() {        
-        return null; // Replace null with the variable you used to store the result
+  }
+
+  /**
+   * Question 24
+   * Retrieves the Opportunity Name, Account Name, and Owner Name for Opportunities that have a Stage
+   *  of 'Closed Won'.
+   * Example: getOpportunityDetailsForClosedWon()
+   * @return A list of filtered Opportunity
+   */
+  public static List<Opportunity> getOpportunityDetailsForClosedWon() {
+    return [
+      SELECT Id, Name, AccountId, Account.Name, Owner.Name, Primary_Contact__c
+      FROM Opportunity
+      WHERE StageName = 'Closed Won'
+    ];
+  }
+
+  /**
+   * Question 25
+   * Sets the 'Primary_Contact__c' field for all Closed Won Opportunities to first Contact on the Account.
+   * Example: setPrimaryContact() should set the 'Primary_Contact__c' field for every Opportunity in the org
+   */
+  public static void setPrimaryContact() {
+    //Use the getOpportunityDetailsForClosedWon() method to get the Opportunities
+    List<Opportunity> closedWonOppLst = SOQL.getOpportunityDetailsForClosedWon();
+    // Create a set of Account Ids from the Opportunities
+    Map<Id, Id> acctToContactIdMap = new Map<Id, Id>();
+    for (Opportunity opp : closedWonOppLst) {
+      acctToContactIdMap.put(opp.AccountId, null);
     }
-
-    /**
-     * Question 25
-     * Sets the 'Primary_Contact__c' field for all Closed Won Opportunities to first Contact on the Account.
-     * Example: setPrimaryContact() should set the 'Primary_Contact__c' field for every Opportunity in the org
-     */
-    public static void setPrimaryContact() {
-        //Use the getOpportunityDetailsForClosedWon() method to get the Opportunities
-
-        // Create a set of Account Ids from the Opportunities
-
-        // Query the Contacts
-
-        // Loop through each Contact and add the Account Id and Contact to the map
-        
-        // Loop through each Opportunity
-
-                // Update the Opportunity primary contact from accountToContactMap
-
-        // Update the Opportunities
+    // Query the Contacts
+    List<Contact> conLst = [
+      SELECT Id, LastName, AccountId
+      FROM Contact
+      WHERE AccountId IN :acctToContactIdMap.keySet()
+      ORDER BY Lastname ASC
+    ];
```

Comment:
It's more advanced and you'll have to rework the entire thing but can you query both the accounts and opportunities at the same time?